File: DAA_Project/heap_range.py
"""Heap-based solution for the smallest range covering at least one element from each sorted array."""

from dataclasses import dataclass
import heapq
from typing import List, Tuple


@dataclass(order=True)
class Node:
    """Represents one element in the heap with its source array position."""

    value: int
    array_index: int
    element_index: int


def find_smallest_range(arrays: List[List[int]], verbose: bool = False) -> Tuple[int, int]:
    """
    Find the smallest range [start, end] that includes at least one number
    from each sorted array.

    Algorithm:
    - Step 1: Initialize heap with the first element from every array.
    - Step 2: Track current maximum among heap elements.
    - Step 3: Repeatedly pop minimum and update best range.
    - Step 4: Push next element from the popped element's array.
    - Step 5: Stop when any array gets exhausted.

    Time Complexity: O(nk log k) where k = number of arrays, n = average array size.
    Space Complexity: O(k)
    """
    if not arrays or any(len(arr) == 0 for arr in arrays):
        raise ValueError("All arrays must be non-empty.")

    # Step 1: Initialize heap
    min_heap: List[Node] = []
    current_max = float("-inf")

    for i, arr in enumerate(arrays):
        node = Node(arr[0], i, 0)
        heapq.heappush(min_heap, node)
        current_max = max(current_max, arr[0])

    best_start = min_heap[0].value
    best_end = int(current_max)

    if verbose:
        print("Initial heap built with first element from each array.")
        print(f"Current range: [{best_start}, {best_end}]")

    while True:
        # Current minimum comes from heap root.
        min_node = heapq.heappop(min_heap)
        current_min = min_node.value

        # Update best range if this one is tighter.
        if current_max - current_min < best_end - best_start:
            best_start, best_end = current_min, int(current_max)
            if verbose:
                print(f"Updated best range -> [{best_start}, {best_end}]")

        next_index = min_node.element_index + 1
        source_array = arrays[min_node.array_index]

        # Stop when any array is exhausted.
        if next_index >= len(source_array):
            break

        next_value = source_array[next_index]
        heapq.heappush(
            min_heap,
            Node(next_value, min_node.array_index, next_index),
        )

        # Step 2: Track max value
        if next_value > current_max:
            current_max = next_value

        if verbose:
            print(
                "Popped min:",
                current_min,
                "| pushed next:",
                next_value,
                "| current max:",
                current_max,
            )

    return best_start, best_end

File: heap_range.c
#include "heap_range.h"

#include <limits.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int value;
    int array_index;
    int element_index;
} HeapNode;

static void swap_nodes(HeapNode *a, HeapNode *b) {
    HeapNode temp = *a;
    *a = *b;
    *b = temp;
}

static void heapify_up(HeapNode *heap, int index) {
    while (index > 0) {
        int parent = (index - 1) / 2;
        if (heap[parent].value <= heap[index].value) {
            break;
        }
        swap_nodes(&heap[parent], &heap[index]);
        index = parent;
    }
}

static void heapify_down(HeapNode *heap, int size, int index) {
    while (1) {
        int left = 2 * index + 1;
        int right = 2 * index + 2;
        int smallest = index;

        if (left < size && heap[left].value < heap[smallest].value) {
            smallest = left;
        }
        if (right < size && heap[right].value < heap[smallest].value) {
            smallest = right;
        }
        if (smallest == index) {
            break;
        }

        swap_nodes(&heap[index], &heap[smallest]);
        index = smallest;
    }
}

static void heap_push(HeapNode *heap, int *size, HeapNode value) {
    heap[*size] = value;
    heapify_up(heap, *size);
    (*size)++;
}

static HeapNode heap_pop(HeapNode *heap, int *size) {
    HeapNode root = heap[0];
    (*size)--;
    heap[0] = heap[*size];
    heapify_down(heap, *size, 0);
    return root;
}

int find_smallest_range(
    const IntArray *arrays,
    int k,
    int verbose,
    int *out_start,
    int *out_end
) {
    int i;
    int heap_size = 0;
    int current_max = INT_MIN;
    int best_start;
    int best_end;
    HeapNode *heap;

    if (arrays == NULL || k <= 0 || out_start == NULL || out_end == NULL) {
        return 0;
    }

    for (i = 0; i < k; i++) {
        if (arrays[i].size <= 0 || arrays[i].data == NULL) {
            return 0;
        }
    }

    heap = (HeapNode *)malloc((size_t)k * sizeof(HeapNode));
    if (heap == NULL) {
        return 0;
    }

    for (i = 0; i < k; i++) {
        HeapNode node;
        node.value = arrays[i].data[0];
        node.array_index = i;
        node.element_index = 0;
        heap_push(heap, &heap_size, node);

        if (node.value > current_max) {
            current_max = node.value;
        }
    }

    best_start = heap[0].value;
    best_end = current_max;

    if (verbose) {
        printf("Initial heap built with first element from each array.\n");
        printf("Current range: [%d, %d]\n", best_start, best_end);
    }

    while (heap_size == k) {
        HeapNode min_node = heap_pop(heap, &heap_size);
        int current_min = min_node.value;
        int next_index = min_node.element_index + 1;
        int array_index = min_node.array_index;

        if ((current_max - current_min) < (best_end - best_start)) {
            best_start = current_min;
            best_end = current_max;
            if (verbose) {
                printf("Updated best range -> [%d, %d]\n", best_start, best_end);
            }
        }

        if (next_index >= arrays[array_index].size) {
            break;
        }

        {
            int next_value = arrays[array_index].data[next_index];
            HeapNode next_node;
            next_node.value = next_value;
            next_node.array_index = array_index;
            next_node.element_index = next_index;
            heap_push(heap, &heap_size, next_node);

            if (next_value > current_max) {
                current_max = next_value;
            }

            if (verbose) {
                printf(
                    "Popped min: %d | pushed next: %d | current max: %d\n",
                    current_min,
                    next_value,
                    current_max
                );
            }
        }
    }

    free(heap);
    *out_start = best_start;
    *out_end = best_end;
    return 1;
}